### PR TITLE
feat(terraform): Dynamic Blocks support for loop in for_each attribute

### PR DESCRIPTION
**Pull request by ChanochShayner**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

## Description
Dynamic Blocks support for loop in for_each attribute.
The previous test was improved according to the implementation.

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/terraform/graph_builder/variable_rendering/renderer.py [hunk 1]
```diff
@@ -398,12 +406,31 @@ def _assign_dynamic_value_for_list(
...
         dynamic_value_in_map = TerraformVariableRenderer.extract_dynamic_value_in_map(
-            dpath.get(block_content, dynamic_argument, separator='.')
+            dpath.get(block_content, dynamic_argument, separator='.'), dynamic_argument
         )
         if block_name not in dynamic_value and dynamic_value_in_map in dynamic_value:
             dpath.set(block_conf, dynamic_argument, dynamic_value[dynamic_value_in_map], separator='.')
         else:
-            dpath.set(block_conf, dynamic_argument, dynamic_value[block_name][0][dynamic_value_in_map], separator='.')
+            try:
+                dpath.set(block_conf, dynamic_argument, dynamic_value[block_name][0][dynamic_value_in_map], separator='.')
```
Comment (rotemavni) [inline checkov/terraform/graph_builder/variable_rendering/renderer.py:415]: `'.'` could also be a const

File: checkov/terraform/graph_builder/variable_rendering/renderer.py [hunk 2]
```diff
@@ -398,12 +406,31 @@ def _assign_dynamic_value_for_list(
...
+            except KeyError:
+                if block_content.get(dynamic_argument) and 'lookup' in block_content.get(dynamic_argument):
```
Comment (rotemavni) [inline checkov/terraform/graph_builder/variable_rendering/renderer.py:417]: `lookup` could be a const too

File: checkov/terraform/graph_builder/variable_rendering/renderer.py [hunk 3]
```diff
@@ -501,3 +525,31 @@ def evaluate_value(self, val: Any) -> Any:
                 evaluated_key = self.evaluate_value(k)
                 evaluated_val[evaluated_key] = self.evaluate_value(v)
         return evaluated_val
+
+
+def find_match_bracket_index(s: str, open_bracket_idx: int) -> int:
+    res = {}
+    pstack = []
+    for i, c in enumerate(s):
+        if c == '[':
```
Comment (rotemavni) [inline checkov/terraform/graph_builder/variable_rendering/renderer.py:534]: don't you have a const for it?

File: checkov/terraform/graph_builder/variable_rendering/renderer.py [hunk 4]
```diff
@@ -501,3 +525,31 @@ def evaluate_value(self, val: Any) -> Any:
...
+            pstack.append(i)
+        elif c == ']':
```
Comment (rotemavni) [inline checkov/terraform/graph_builder/variable_rendering/renderer.py:536]: same here

**Review by rotemavni — APPROVED** [review]
🦾

File: checkov/terraform/graph_builder/variable_rendering/renderer.py [hunk 5]
```diff
@@ -501,3 +527,31 @@ def evaluate_value(self, val: Any) -> Any:
...
+
+def find_match_bracket_index(s: str, open_bracket_idx: int) -> int:
+    res = {}
+    pstack = []
+    for i, c in enumerate(s):
+        if c == LEFT_BRACKET:
+            pstack.append(i)
+        elif c == RIGHT_BRACKET:
+            if len(pstack) == 0:
+                raise IndexError("No matching closing parens at: " + str(i))
```
Comment (gruebel) [inline checkov/terraform/graph_builder/variable_rendering/renderer.py:540]:
```suggestion
                raise IndexError("No matching closing brackets at: " + str(i))
```

**Pull request merged**